internal/rbd/controllerserver.go

```diff
@@ -1701,3 +1701,27 @@ func (cs *ControllerServer) ControllerUnpublishVolume(
 
 	return &csi.ControllerUnpublishVolumeResponse{}, nil
 }
+
+func (cs *ControllerServer) GetSnapshot(
+	ctx context.Context,
+	req *csi.GetSnapshotRequest,
+) (*csi.GetSnapshotResponse, error) {
+	mgr := NewManager(cs.Driver.GetInstanceID(), nil, req.GetSecrets())
+	defer mgr.Destroy(ctx)
+
+	snapID := req.GetSnapshotId()
+	snapshot, err := mgr.GetSnapshotByID(ctx, snapID)
+	if err != nil {
+		return nil, status.Errorf(codes.NotFound, "failed to find snapshot with ID %q: %s", snapID, err.Error())
+	}
+	defer snapshot.Destroy(ctx)
+
+	csiSnap, err := snapshot.ToCSI(ctx)
```

Comment:
`ToCSI` will fail due to missing `SourceVolumeID` on the fetched snapshot. However, we can get the same from the incoming request.

Reply:
Indeed, that is interesting! `genSnapFromSnapID()` should return a complete `rbdSnapshot` struct, but this detail is missing. I wonder if it is intentionally not stored/fetched in the journal...

Will look into this a little more soon.